File: src/Infrastructure/Finance/FinanceService.cs
using ConstructionManagement.Application.Finance;
using ConstructionManagement.Domain.Entities;
using ConstructionManagement.Infrastructure.Persistence;
using Microsoft.EntityFrameworkCore;

namespace ConstructionManagement.Infrastructure.Finance;

public sealed class FinanceService(ConstructionDbContext dbContext) : IFinanceService
{
    public async Task<IReadOnlyList<ExpenseListItem>> GetExpensesAsync(CancellationToken cancellationToken = default)
    {
        return await dbContext.Expenses.AsNoTracking().OrderByDescending(expense => expense.Date)
            .Select(expense => ToListItem(expense)).ToListAsync(cancellationToken);
    }

    public async Task<ExpenseListItem> CreateExpenseAsync(CreateExpenseRequest request, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(request.Description)) throw new ArgumentException("Expense description is required.");
        if (request.Amount <= 0) throw new ArgumentException("Expense amount must be greater than zero.");

        var expense = new Expense
        {
            ProjectId = request.ProjectId,
            PhaseId = request.PhaseId,
            Category = request.Category,
            Amount = request.Amount,
            Date = request.Date,
            Vendor = request.Vendor,
            Description = request.Description,
            PaymentMethod = request.PaymentMethod,
        };
        dbContext.Expenses.Add(expense);
        await dbContext.SaveChangesAsync(cancellationToken);
        return ToListItem(expense);
    }

    public async Task<IReadOnlyList<PaymentListItem>> GetPaymentsAsync(CancellationToken cancellationToken = default)
    {
        return await dbContext.Payments.AsNoTracking().OrderByDescending(payment => payment.Date)
            .Select(payment => ToListItem(payment)).ToListAsync(cancellationToken);
    }

    public async Task<PaymentListItem> CreatePaymentAsync(CreatePaymentRequest request, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(request.InvoiceNumber)) throw new ArgumentException("Invoice or reference is required.");
        if (request.Amount <= 0) throw new ArgumentException("Payment amount must be greater than zero.");
        if (await dbContext.Payments.AnyAsync(payment => payment.InvoiceNumber == request.InvoiceNumber, cancellationToken))
            throw new InvalidOperationException("An invoice or reference with this number already exists.");

        var payment = new Payment
        {
            ProjectId = request.ProjectId,
            PhaseId = request.PhaseId,
            PaymentType = request.PaymentType,
            PartyName = request.PartyName,
            InvoiceNumber = request.InvoiceNumber,
            Amount = request.Amount,
            Date = DateTime.UtcNow,
            DueDate = request.DueDate,
            Status = request.Status,
        };
        dbContext.Payments.Add(payment);
        await dbContext.SaveChangesAsync(cancellationToken);
        return ToListItem(payment);
    }

    private static ExpenseListItem ToListItem(Expense expense) => new(
        expense.Id, expense.ProjectId, expense.PhaseId, expense.Category, expense.Amount,
        expense.Date, expense.Vendor, expense.Description, expense.PaymentMethod);

    private static PaymentListItem ToListItem(Payment payment) => new(
        payment.Id, payment.ProjectId, payment.PhaseId, payment.PaymentType, payment.PartyName,
        payment.InvoiceNumber, payment.Amount, payment.Date, payment.DueDate, payment.Status);
}

File: src/Web/src/app/finance.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, catchError, of } from 'rxjs';

export interface ApiExpense {
  id: string;
  projectId: string;
  phaseId: string | null;
  category: string;
  amount: number;
  date: string;
  vendor: string;
  description: string;
  paymentMethod: string;
}

export interface ApiPayment {
  id: string;
  projectId: string;
  phaseId: string | null;
  paymentType: string;
  partyName: string;
  invoiceNumber: string;
  amount: number;
  date: string;
  dueDate: string | null;
  status: string;
}

@Injectable({ providedIn: 'root' })
export class FinanceService {
  private readonly http = inject(HttpClient);
  private readonly expensesEndpoint = '/api/expenses';
  private readonly paymentsEndpoint = '/api/payments';

  getExpenses(): Observable<ApiExpense[]> {
    return this.http.get<ApiExpense[]>(this.expensesEndpoint).pipe(catchError(() => of([])));
  }

  createExpense(request: {
    projectId: string;
    phaseId: string;
    category: string;
    amount: number;
    date: string;
    vendor: string;
    description: string;
    paymentMethod: string;
  }): Observable<ApiExpense> {
    return this.http.post<ApiExpense>(this.expensesEndpoint, request);
  }

  getPayments(): Observable<ApiPayment[]> {
    return this.http.get<ApiPayment[]>(this.paymentsEndpoint).pipe(catchError(() => of([])));
  }

  createPayment(request: {
    projectId: string;
    phaseId: string;
    paymentType: string;
    partyName: string;
    invoiceNumber: string;
    amount: number;
    dueDate: string | null;
    status: string;
  }): Observable<ApiPayment> {
    return this.http.post<ApiPayment>(this.paymentsEndpoint, request);
  }
}
